x/airdrop/keeper/claim_action.go

```diff
@@ -0,0 +1,138 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	gov "github.com/cosmos/cosmos-sdk/x/gov/types"
+
+	"github.com/ingenuity-build/quicksilver/x/airdrop/types"
+)
+
+// VerifyClaimAction validates the claim against the specific action
+// conditions. It is the responsibility of the caller to ensure that the given
+// action is in bounds by calling action.InBounds().
+//
+// TODO: we also want to verify that the action was executed on the remote
+// chain before we execute the claim...
+func (k Keeper) VerifyClaimAction(ctx sdk.Context, cr types.ClaimRecord, action types.Action) error {
+	// action already completed, nothing to claim
+	if _, exists := cr.ActionsCompleted[int32(action)]; exists {
+		return fmt.Errorf("%s already completed", types.Action_name[int32(action)])
+	}
+
+	switch action {
+	case types.ActionInitialClaim:
+		return nil
+	case types.ActionDepositT1:
+		return k.checkDeposit(ctx, cr, sdk.MustNewDecFromStr("0.05"))
+	case types.ActionDepositT2:
+		return k.checkDeposit(ctx, cr, sdk.MustNewDecFromStr("0.10"))
+	case types.ActionDepositT3:
+		return k.checkDeposit(ctx, cr, sdk.MustNewDecFromStr("0.15"))
+	case types.ActionDepositT4:
+		return k.checkDeposit(ctx, cr, sdk.MustNewDecFromStr("0.22"))
+	case types.ActionDepositT5:
+		return k.checkDeposit(ctx, cr, sdk.MustNewDecFromStr("0.30"))
+	case types.ActionStakeQCK:
+		return k.checkBondedDelegation(ctx, cr.Address)
+	case types.ActionSignalIntent:
+		return k.checkIntentIsSet(ctx, cr.ChainId, cr.Address)
+	case types.ActionQSGov:
+		return k.checkQSGov(ctx, cr.Address)
+	case types.ActionGbP:
+		// TODO: implement check once GbP is implemented
+	case types.ActionOsmosis:
+		// IBC proof based verification on Osmosis remote zone
+		// TODO: implement
+	default:
+		return fmt.Errorf("undefined action [%d]", action)
+	}
+
+	return fmt.Errorf("verification not implemented for [%d] %s", action, types.Action_name[int32(action)])
+}
+
+// checkDeposit checks
+func (k Keeper) checkDeposit(ctx sdk.Context, cr types.ClaimRecord, threshold sdk.Dec) error {
+	addr, err := sdk.AccAddressFromBech32(cr.Address)
+	if err != nil {
+		return err
+	}
+
+	zone, ok := k.icsKeeper.GetZone(ctx, cr.ChainId)
+	if !ok {
+		return fmt.Errorf("zone not found for %s", cr.ChainId)
+	}
+
+	dAmount := sdk.NewCoin(
+		zone.BaseDenom,
+		threshold.MulInt64(int64(cr.BaseValue)).TruncateInt(),
+	)
+
+	if !k.bankKeeper.HasBalance(ctx, addr, dAmount) {
```

Comment:
We want this to succeed even if users have sent their tokens to Osmosis, for example.

This needs to be a lookup against receipts for that (user|zone) tuple.

We added an iterator on receipts for user/zone, so we should trivially be able to add a function to fetch the summed value of deposits for a given zone.

Reply:
are we looking for the net deposit amount (deposits - redemptions) or simply gross deposits?

Reply:
net deposit is fine